File: ts/mergeIntervals.ts
function mergeIntervals(intervals: number[][]): number[][] {
  intervals.sort((a, b) => a[0] - b[0]);
  const result: number[][] = [];
  const first = intervals.at(0) as number[];
  let min: number = first[0];
  let max: number = first[1];

  for (let i = 1; i < intervals.length; i++) {
    const [start, end] = intervals[i];

    if (max < start) {
      result.push([min, max]);

      min = start;
      max = end;
    } else {
      min = Math.min(min, start);
      max = Math.max(max, end);
    }
  }

  result.push([min, max]);

  return result;
}

export default () => {
  const intervals = [
    [1, 4],
    [0, 0],
  ];
  // const intervals = [
  //   [1, 4],
  //   [0, 4],
  // ];
  // const intervals = [
  //   [1, 4],
  //   [4, 5],
  // ];
  // const intervals = [
  //   [1, 3],
  //   [2, 6],
  //   [8, 10],
  //   [15, 18],
  // ];
  const result = mergeIntervals(intervals);
  console.log(result);
};

File: rs/src/merge_intervals.rs
fn merge_intervals(mut intervals: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
    intervals.sort();
    let mut result: Vec<Vec<i32>> = vec![];
    let first = &intervals[0];
    let mut min = first[0];
    let mut max = first[1];

    for interval in intervals.iter().skip(1) {
        let start = interval[0];
        let end = interval[1];

        if max < start {
            result.push(vec![min, max]);

            min = start;
            max = end;
        } else {
            min = min.min(start);
            max = max.max(end);
        }
    }

    result.push(vec![min, max]);

    return result;
}

pub fn main() {
    let intervals = vec![vec![1, 3], vec![2, 6], vec![8, 10], vec![15, 18]];

    let result = merge_intervals(intervals);
    println!("{:?}", result);
}
